### PR TITLE
fix(mcp): keep connection cleanup in owner tasks

### LINKED ISSUE #353: MCP cleanup exits AnyIO cancel scopes from a different asyncio task
## Summary

`MCPManager` keeps MCP stdio and session contexts in one shared `AsyncExitStack`. A connection may be opened from a parallel-agent or for-each worker task, while provider shutdown later closes that stack from the workflow/root task.

AnyIO cancel scopes are task-affine, so cleanup can fail with:

```text
RuntimeError: Attempted to exit cancel scope in a different task than it was entered in
```

The exception is currently caught and logged by `MCPManager.close()`, which means shutdown continues but MCP sessions/processes may not be cleaned up correctly.

## Public `main` references

Verified against `origin/main` at `c8fad5c95fdae0e748455041ed1eca706329989d`:

- The manager owns one shared `AsyncExitStack`: https://github.com/microsoft/conductor/blob/c8fad5c95fdae0e748455041ed1eca706329989d/src/conductor/mcp/manager.py#L122-L127
- `connect_server()` enters both AnyIO-backed contexts in the calling task: https://github.com/microsoft/conductor/blob/c8fad5c95fdae0e748455041ed1eca706329989d/src/conductor/mcp/manager.py#L177-L186
- `close()` exits the stack in whichever task performs provider cleanup: https://github.com/microsoft/conductor/blob/c8fad5c95fdae0e748455041ed1eca706329989d/src/conductor/mcp/manager.py#L494-L508
- Claude pools managers used by concurrent agents: https://github.com/microsoft/conductor/blob/c8fad5c95fdae0e748455041ed1eca706329989d/src/conductor/providers/claude.py#L627-L634
- Provider shutdown closes those managers later: https://github.com/microsoft/conductor/blob/c8fad5c95fdae0e748455041ed1eca706329989d/src/conductor/providers/claude.py#L755-L770

## Minimal reproduction of the lifecycle pattern

This does not require an API call or Pydantic AI; it isolates the same task-ownership mismatch:

```python
import asyncio
from contextlib import AsyncExitStack, asynccontextmanager

import anyio


@asynccontextmanager
async def task_bound_context():
    with anyio.CancelScope():
        yield


async def main():
    stack = AsyncExitStack()

    async def connect():
        await stack.enter_async_context(task_bound_context())

    await asyncio.create_task(connect())
    await stack.aclose()


asyncio.run(main())
```

Observed with Python 3.12.13 and AnyIO 4.12.1:

```text
RuntimeError: Attempted to exit cancel scope in a different task than it was entered in
```

In Conductor, parallel and for-each execution provide the worker-task side of this pattern; provider/registry cleanup provides the root-task side.

## Expected behavior

MCP connections should close cleanly regardless of which workflow task first requested the manager, including parallel and for-each runs. No task-affinity exception should be logged, and all sessions/transports/processes should be released.

## Suggested direction

Give each `MCPManager` a persistent lifecycle-owner task and marshal both context entry and context exit through that same task. Regression coverage should include:

1. connect from a worker task, close from its parent task;
2. failed connection followed by close;
3. cancelled connection followed by close;
4. multiple pooled managers created by parallel agents.

Closing only from the original caller is insufficient because that task may already have completed by provider shutdown.

### PR DESCRIPTION
**Pull request by hertznsk**
## Summary

- move each MCP stdio/session context lifecycle into a persistent owner task
- signal owner tasks during manager shutdown so AnyIO cancel scopes exit in the task that entered them
- preserve caller cancellation while consuming and logging cleanup failures
- protect owner-task teardown when manager shutdown is cancelled
- reject duplicate server names without orphaning the existing connection

## Testing

- `PYTHONASYNCIODEBUG=1 uv run pytest tests/test_mcp -q` (51 passed)
- `make check`
- direct stdio MCP lifecycle driver covering connect, tool call, duplicate names, failure, cancellation, concurrency, and idempotent close

Fixes #353

### CODE REVIEW

File: src/conductor/mcp/manager.py [hunk 1]
```diff
@@ -491,25 +539,52 @@ def has_servers(self) -> bool:
...
+                        logger.warning(f"Error closing MCP connections: {result}")
+                raise
 
-        self.sessions.clear()
-        self.tools.clear()
-        self.tool_to_server.clear()
-        self._initialized = False
+            for result in results:
+                if isinstance(result, Exception):
+                    logger.warning(f"Error closing MCP connections: {result}")
```
Comment (jrob5756) [inline src/conductor/mcp/manager.py]: If close() gets cancelled a second time while it's sitting at the unshielded `await cleanup` on line 573, that cancellation propagates into the still-running owner tasks and can interrupt teardown mid-flight (e.g. while a subprocess is being terminated). But `cleanup.done()` is true even for a cancelled future, so this `finally` block clears all bookkeeping as if everything shut down cleanly, and there's no log line for this path at all. Worth tracking whether `results` was actually obtained before treating state as cleared, and logging when a repeated cancellation prevents confirming cleanup completed.

File: src/conductor/mcp/manager.py [hunk 2]
```diff
@@ -163,59 +165,105 @@ async def connect_server(
...
+        try:
+            tools = await asyncio.shield(ready)
+        except asyncio.CancelledError:
+            if not ready.done():
+                ready.cancel()
+            task.cancel()
+            cleanup = asyncio.gather(task, return_exceptions=True)
+            try:
+                results = await asyncio.shield(cleanup)
+            except asyncio.CancelledError:
```
Comment (jrob5756) [inline src/conductor/mcp/manager.py]: Same gap here as in close(): if a second cancellation lands while awaiting the unshielded `await cleanup` on this line, it propagates straight out and skips the pop()/`_discard_server_state()` calls below. That leaves `_connection_tasks[name]` populated with a task that may not have finished cleanup, so subsequent `connect_server(name=...)` calls for that name will incorrectly raise "already connected or connecting" until the whole manager is torn down, with no log explaining why.

File: src/conductor/mcp/manager.py [hunk 3]
```diff
@@ -163,59 +165,105 @@ async def connect_server(
...
+            self._connection_tasks.pop(name, None)
+            self._connection_stops.pop(name, None)
+            self._discard_server_state(name)
+            raise
+        except Exception as exc:
+            await asyncio.gather(task, return_exceptions=True)
+            self._connection_tasks.pop(name, None)
+            self._connection_stops.pop(name, None)
+            self._discard_server_state(name)
+            logger.error(f"Failed to connect to MCP server '{name}': {exc}")
```
Comment (jrob5756) [inline src/conductor/mcp/manager.py]: `{exc}` only calls `str()`. If the `AsyncExitStack` teardown raises a different exception while unwinding after the real connect failure (e.g. a broken pipe during `stdio_client`/`ClientSession` exit), that secondary exception replaces the original as what's caught and logged here, masking the actual root cause.

```suggestion
            logger.error(f"Failed to connect to MCP server '{name}': {exc}", exc_info=exc)
```

File: src/conductor/mcp/manager.py [hunk 4]
```diff
@@ -163,59 +165,105 @@ async def connect_server(
...
+            if not ready.done():
+                ready.cancel()
+            task.cancel()
+            cleanup = asyncio.gather(task, return_exceptions=True)
+            try:
+                results = await asyncio.shield(cleanup)
+            except asyncio.CancelledError:
+                results = await cleanup
+            for result in results:
+                if isinstance(result, Exception) and not isinstance(result, asyncio.CancelledError):
```
Comment (jrob5756) [inline src/conductor/mcp/manager.py]: `asyncio.CancelledError` derives from `BaseException`, not `Exception`, since Python 3.8, so `isinstance(result, Exception)` already excludes it. The trailing check is dead code.

```suggestion
                if isinstance(result, Exception):
```

File: src/conductor/mcp/manager.py [hunk 5]
```diff
@@ -119,10 +120,11 @@ def __init__(self, tool_output: ToolOutputConfig | None = None) -> None:
...
             raise ImportError("MCP SDK not installed. Install with: uv add 'mcp>=1.0.0'")
 
-        self.sessions: dict[str, ClientSession] = {}
+        self.sessions: dict[str, Any] = {}
         self.tools: dict[str, list[dict[str, Any]]] = {}  # server -> tools
         self.tool_to_server: dict[str, str] = {}  # prefixed_name -> server
-        self._exit_stack = AsyncExitStack()
+        self._connection_tasks: dict[str, asyncio.Task[None]] = {}
+        self._connection_stops: dict[str, asyncio.Event] = {}
         self._initialized = False
```
Comment (jrob5756) [inline src/conductor/mcp/manager.py]: `_initialized` is set here and at two other spots but nothing reads it anymore — `close()`'s guard now checks `self._connection_tasks` instead. Worth removing it so a future reader doesn't assume it reflects live connection state.

File: src/conductor/mcp/manager.py [hunk 6]
```diff
@@ -491,25 +539,52 @@ def has_servers(self) -> bool:
...
+
     async def close(self) -> None:
         """Close all server connections and clean up resources.
 
         This method should be called when the manager is no longer needed.
         It properly closes all stdio connections and cleans up internal state.
         """
-        if not self._initialized:
+        if not self._connection_tasks:
             return
```
Comment (jrob5756) [inline src/conductor/mcp/manager.py:568]: This counts `len(self.sessions)`, which is only populated once a connection's `ready` future resolves. If `close()` runs while a `connect_server()` call is still mid-handshake, this understates how many owner tasks are actually being torn down. `len(self._connection_tasks)` would be accurate.

**Review by jrob5756 — COMMENTED** [review]
Good fix overall. The owner-task-per-connection model is the right way to solve the anyio cancel-scope task affinity problem, and the new task-affinity tests are a solid way to prove it. 

Left a few small comments on double-cancellation edge cases and a couple of small cleanups, but nothing blocking here.

**Review by hertznsk — COMMENTED** [review]
All six review comments are addressed in ca8551d. Per-thread replies via the review-comment API return 404 for my token (fork-based PR), so summarizing here:

1. close() double-cancellation — the unshielded `await cleanup` fallback is replaced with a re-shielding loop: repeated cancellations land on a throwaway `asyncio.shield(cleanup)` while the gather keeps running, and bookkeeping is cleared only after `results` is actually obtained. close() now deliberately absorbs cancellation until owner-task teardown completes (documented in the docstring); covered by `test_double_cancelled_close_still_finishes_cleanup`.
2. connect_server() double-cancellation — same re-shielding loop: cleanup completes and the `pop()`/`_discard_server_state()` calls always run before the cancellation is re-raised, so `_connection_tasks[name]` can no longer be left populated and block reconnects; covered by `test_double_cancelled_connection_clears_bookkeeping`.
3. Connect-failure logging — now passes `exc_info=exc`, so a secondary AsyncExitStack teardown exception cannot mask the original root cause.
4. Dead `not isinstance(result, asyncio.CancelledError)` check removed.
5. Unused `_initialized` flag removed from all three spots.
6. Shutdown log now counts `len(self._connection_tasks)`, including mid-handshake connections.

Full test suite passes (4580 tests), ruff and ty clean.

File: src/conductor/mcp/manager.py (hunk 6 above)
Comment (hertznsk) [reply src/conductor/mcp/manager.py:568]: Fixed in ca8551d - the debug log now counts len(self._connection_tasks), which includes connections still mid-handshake.

File: src/conductor/mcp/manager.py (hunk 1 above)
Comment (hertznsk) [reply src/conductor/mcp/manager.py]: Fixed in ca8551d. The unshielded `await cleanup` fallback is replaced with a re-shielding loop: repeated cancellations now land on a throwaway `asyncio.shield(cleanup)` while the gather keeps running, and bookkeeping is cleared only after `results` is actually obtained (a cancelled gather result is never read). The behavior goes slightly further than suggested: close() now deliberately absorbs cancellation until owner-task teardown completes (documented in the docstring), so teardown is never interrupted mid-flight and state is always consistent. Covered by test_double_cancelled_close_still_finishes_cleanup.

File: src/conductor/mcp/manager.py (hunk 2 above)
Comment (hertznsk) [reply src/conductor/mcp/manager.py]: Fixed in ca8551d with the same re-shielding loop as in close(): a repeated cancellation lands on the shield instead of the gather, so cleanup completes and the pop()/_discard_server_state() calls always run before the cancellation is re-raised. _connection_tasks[name] can no longer be left populated, so a later connect_server(name=...) will not be spuriously rejected. Covered by test_double_cancelled_connection_clears_bookkeeping.

File: src/conductor/mcp/manager.py (hunk 3 above)
Comment (hertznsk) [reply src/conductor/mcp/manager.py]: Applied in ca8551d - the log call now passes exc_info=exc, so the original connect failure traceback is preserved even when AsyncExitStack teardown raises a secondary exception while unwinding.

File: src/conductor/mcp/manager.py (hunk 4 above)
Comment (hertznsk) [reply src/conductor/mcp/manager.py]: Applied in ca8551d - the dead CancelledError check is removed; the condition is now plain isinstance(result, Exception).

File: src/conductor/mcp/manager.py (hunk 5 above)
Comment (hertznsk) [reply src/conductor/mcp/manager.py]: Removed in ca8551d - _initialized is gone from all three spots; close() keeps guarding on _connection_tasks.

**Review by jrob5756 — APPROVED** [review]
LGTM. Approved!
